braille-printer-app/brf-printer-app.c

```diff
@@ -0,0 +1,821 @@
+//
+// Local functions...
+//
+static bool BRFTestFilterCB(pappl_job_t *job,  pappl_device_t *device,void *cbdata) ; 
+static int brf_print_filter_function(int inputfd,int outputfd, int inputseekable,cf_filter_data_t *data, void *parameters); 
+static const char *autoadd_cb(const char *device_info, const char *device_uri, const char *device_id, void *cbdata);
+static bool	driver_cb(pappl_system_t *system, const char *driver_name, const char *device_uri, const char *device_id, pappl_pr_driver_data_t *data, ipp_t **attrs, void *cbdata);
+static int	match_id(int num_did, cups_option_t *did, const char *match_id);
+static const char *mime_cb(const unsigned char *header, size_t headersize, void *data);
+static bool	printer_cb(const char *device_info, const char *device_uri, const char *device_id, pappl_system_t *system);
+static pappl_system_t *system_cb(int num_options, cups_option_t *options, void *data);
+
+
+//
+// Local globals...
+//
+
+static pappl_pr_driver_t	brf_drivers[] =
+{					// Driver list
+{ "gen_brf",  "Generic",
+  NULL, NULL },
+
+};
+static char			brf_statefile[1024];
+					// State file
+
+
+//
+// 'main()' - Main entry for brf.
+//
+
+int					// O - Exit status
+main(int  argc,				// I - Number of command-line arguments
+     char *argv[])			// I - Command-line arguments
+{
+  return (papplMainloop(argc, argv,
+                        "1.0",
+                        NULL,
+                        (int)(sizeof(brf_drivers) / sizeof(brf_drivers[0])),
+                        brf_drivers, autoadd_cb, driver_cb,
+                        /*subcmd_name*/NULL, /*subcmd_cb*/NULL,
+                        system_cb,
+                        /*usage_cb*/NULL,
+                        /*data*/NULL));
+}
+
+
+//
+// 'autoadd_cb()' - Determine the proper driver for a given printer.
+//
+
+static const char *			// O - Driver name or `NULL` for none
+autoadd_cb(const char *device_info,	// I - Device information/name (not used)
+           const char *device_uri,	// I - Device URI
+           const char *device_id,	// I - IEEE-1284 device ID
+           void       *cbdata)		// I - Callback data (System)
+{
+  int		i,			// Looping var
+		score,			// Current driver match score
+	    	best_score = 0,		// Best score
+		num_did;		// Number of device ID key/value pairs
+  cups_option_t	*did;			// Device ID key/value pairs
+  const char	*make,			// Manufacturer name
+		*best_name = NULL;	// Best driver
+ 
+
+
+  (void)device_info;
+
+  // First parse the device ID and get any potential driver name to match...
+  num_did = papplDeviceParseID(device_id, &did);
+
+  if ((make = cupsGetOption("MANUFACTURER", num_did, did)) == NULL)
+    if ((make = cupsGetOption("MANU", num_did, did)) == NULL)
+      make = cupsGetOption("MFG", num_did, did);
+
+  
+  // // Then loop through the driver list to find the best match...
+  // for (i = 0; i < (int)(sizeof(brf_drivers) / sizeof(brf_drivers[0])); i ++)
+  {
+ 
+
+    if (brf_drivers[i].device_id)
+    {
```

Comment:
brf_drivers[] has only 1 entry without device ID, so `autoadd_cb()` always returns NULL. Did you test the code? Or only compile it?